node_modules/oae-principals/lib/internal/dao.js

```diff
@@ -422,24 +426,109 @@ var setAdmin = module.exports.setAdmin = function(adminType, isAdmin, userId, ca
 };
 
 /**
- * Get a user by their email address. If no user could be found for a given email address,
- * a 404 error will be returned
+ * Get a set of users by their email address
+ *
+ * @param  {String}     email               The email address that is associated with users
+ * @param  {Function}   callback            Standard callback function
+ * @param  {Object}     callback.err        An error object, if any
+ * @param  {User[]}     callback.users      The users that are associated with the email address
+ */
+var getUsersByEmail = module.exports.getUsersByEmail = function(email, callback) {
+    Cassandra.runQuery('SELECT * FROM "PrincipalsByEmail" WHERE "email" = ?', [email], function(err, rows) {
+        if (err) {
+            return callback(err);
+        }
+
+        var userIds = _.map(rows, function(row) {
+            return row.get('principalId').value;
+        });
+        return getPrincipals(userIds, null, callback);
+    });
+};
+
+/**
+ * Get a user via an email token. If no user could be found
+ * for a given token, a 404 error will be returned
  *
- * @param  {String}     email           The email address that is associated with a user
+ * @param  {String}     token           The token that is associated with a user
  * @param  {Function}   callback        Standard callback function
  * @param  {Object}     callback.err    An error object, if any
- * @param  {User}       callback.user   The user that was associated with the email
+ * @param  {User}       callback.user   The user that was associated with the token
  */
-var getUserByEmail = module.exports.getUserByEmail = function(email, callback) {
-    Cassandra.runQuery('SELECT * FROM "PrincipalsByEmail" WHERE "email" = ?', [email], function(err, rows) {
+var getUserByEmailToken = module.exports.getUserByEmailToken = function(token, callback) {
+    Cassandra.runQuery('SELECT * FROM "EmailTokens" WHERE "token" = ?', [token], function(err, rows) {
         if (err) {
             return callback(err);
         } else if (_.isEmpty(rows)) {
-            return callback({'code': 404, 'msg': 'No user found by that email'});
+            return callback({'code': 404, 'msg': 'No user found by that token'});
+        }
+
+        var userId = rows[0].get('userId').value;
+        return getPrincipal(userId, callback);
+    });
+};
+
+/**
+ * Store an email token
+ *
+ * @param  {String}     token           The token to store
+ * @param  {string}     userId          The id of the user with whom the token is associated
+ * @param  {Function}   callback        Standard callback function
+ * @param  {Object}     callback.err    An error object, if any
+ */
+var storeEmailToken = module.exports.storeEmailToken = function(token, userId, callback) {
+    var q = Cassandra.constructUpsertCQL('EmailTokens', 'token', token, {'userId': userId});
```

Comment:
Why are you using `constructUpsertCQL` here, but writing out a query for the other inserts?

Also, can't recall if this was ever supposed to expire? It probably should in case the user put in the incorrect email address.